samples/face_reid/index_builder.py

```diff
@@ -0,0 +1,102 @@
+
+        :param buffer: Gstreamer buffer with this frame's data.
+        :param frame_meta: This frame's metadata.
+        """
+        # image files source adapter adds location tag
+        # which contains image file name
+        location = frame_meta.tags['location']
+
+        if frame_meta.objects_number < 2:
+            # no faces detected and
+            # only primary object is present
+            self.logger.warn(
+                '%s faces detected on %s, 1 is expected. Not adding features to index.',
+                frame_meta.objects_number - 1,
+                location,
+            )
+        else:
+            stream = self.get_cuda_stream(frame_meta)
+            with nvds_to_gpu_mat(buffer, frame_meta.frame_meta) as frame_mat:
+                with Artist(frame_mat, stream) as artist:
+                    root, _ = os.path.splitext(location)
+                    file_name = os.path.basename(root)
+                    # gallery images are named as <person_name>_<img_n>.jpeg
+                    person_name, img_n = file_name.rsplit('_', maxsplit=1)
+                    # assign person_id to person_name in the order of appearance
+                    try:
+                        person_id = self.person_names.index(person_name)
+                    except ValueError:
+                        person_id = len(self.person_names)
+                        self.person_names.append(person_name)
+
+                    img_n = int(img_n)
+                    for obj_meta in frame_meta.objects:
+                        if obj_meta.label == 'face':
+                            # get face feature vector from metadata
+                            feature = obj_meta.get_attr_meta(
+                                REID_MODEL_NAME, 'feature'
+                            ).value
+                            # index stores single int label for each feature
+                            # we assume that there are no more than 2^32 people or images for each person
+                            # and pack both person_id and img_n into single int
+                            feature_id = pack_person_id_img_n(person_id, img_n)
+                            self.index.add_items(feature, feature_id)
+
+                            # save face image to disk
+                            # it will be used to visualize face matches to gallery
+                            face_img = artist.copy_frame_region(
+                                obj_meta.bbox
+                            ).download()
+                            face_img = cv2.cvtColor(face_img, cv2.COLOR_RGBA2BGR)
+                            img_path = os.path.join(
+                                self.processed_gallery_dir,
+                                f'{person_name}_{person_id:03d}_{img_n:03d}.jpeg',
+                            )
+                            cv2.imwrite(img_path, face_img)
+            # refresh index on disk
+            self.write_index()
```

Comment:
Discussed this, decided to leave as is, because there are no obvious points where this save might go.

Added a log message when writing index to disk.